buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -0,0 +1,1428 @@
+            Magazine[] mags;
+            int expansions = 0;
+            do {
+                mags = magazines;
+                int mask = mags.length - 1;
+                int index = (int) (threadId & mask);
+                for (int i = 0, m = Integer.numberOfTrailingZeros(~mask); i < m; i++) {
+                    Magazine mag = mags[index + i & mask];
+                    long writeLock = mag.tryWriteLock();
+                    if (writeLock != 0) {
+                        try {
+                            return mag.allocate(size, sizeBucket, maxCapacity, buf);
+                        } finally {
+                            mag.unlockWrite(writeLock);
+                        }
+                    }
+                }
+                expansions++;
+            } while (expansions <= 3 && tryExpandMagazines(mags.length));
+            // Return the buffer we pulled from the recycler but didn't use.
+            buf.release();
+        }
+        // The magazines failed us, or the buffer is too big to be pooled.
+        return chunkAllocator.allocate(size, maxCapacity);
+    }
+
+    public long usedMemory() {
+        long sum = 0;
+        for (ByteBuf byteBuf : centralQueue) {
+            sum += byteBuf.capacity();
+        }
+        for (Magazine magazine : magazines) {
+            sum += magazine.usedMemory.get();
+        }
+        return sum;
+    }
+
+    private boolean tryExpandMagazines(int currentLength) {
+        if (currentLength >= MAX_STRIPES) {
+            return true;
+        }
+        long writeLock = magazineExpandLock.tryWriteLock();
+        if (writeLock != 0) {
+            try {
+                Magazine[] mags = magazines;
+                if (mags.length >= MAX_STRIPES || mags.length > currentLength) {
+                    return true;
+                }
+                Magazine[] expanded = Arrays.copyOf(mags, mags.length * 2);
+                for (int i = mags.length, m = expanded.length; i < m; i++) {
+                    expanded[i] = new Magazine(this);
+                }
+                magazines = expanded;
+            } finally {
+                magazineExpandLock.unlockWrite(writeLock);
+            }
+        }
+        return true;
+    }
+
+    public void close() {
+        closed = true;
+        long magsExpandWriteLock = magazineExpandLock.writeLock();
+        try {
+            for (Magazine mag : magazines) {
+                long writeLock = mag.writeLock();
+                try {
+                    mag.close();
+                } finally {
+                    mag.unlockWrite(writeLock);
+                }
+            }
+        } finally {
+            magazineExpandLock.unlockWrite(magsExpandWriteLock);
+        }
+        drainCloseCentralQueue();
+    }
+
+    private void drainCloseCentralQueue() {
+        ByteBuf curr;
+        while ((curr = centralQueue.poll()) != null) {
+            curr.release();
```

Comment:
In wonder if we should guard against `IllegalReferenceException` here and so ensure we always poll and release everything even if an exception was thrown. In this case we should rethrow any exceptions once we drained the queue.